### PR TITLE
feat(gatsby): add support for dynamic routes for API functions

### PR DESCRIPTION
**Pull request by KyleAMathews**
Add support for dynamic routes similar to client-side routes e.g. `/src/api/users/[id].js` or `/src/api/products/[...].js`

Extends the tests to support `gatsby serve`.

Lots of refactorings of internal data structures + add an interface.

You can test this PR by installing `gatsby@functions-next`

## TODOs
- [x] move shared util `getMatchPath` to gatsby-core-utils
- [x] more tests
- [x] loop only until find match
- [x] implement in `gatsby serve`
- [x] document matching algorithm (and validate it's the same as in browser)
- [x] copy `manifest.json` to `public/_functions.json` in `gatsby-plugin-google-cloud`
- [x] consolidate data types

### CODE REVIEW

File: packages/gatsby/src/commands/serve.ts [hunk 1]
```diff
@@ -112,17 +113,22 @@ module.exports = async (program: IServeProgram): Promise<void> => {
   const matchPaths = await readMatchPaths(program)
   router.use(matchPathRouter(matchPaths, { root }))
 
-  if (process.env.GATSBY_EXPERIMENTAL_FUNCTIONS) {
```
Comment (dmccraw) [inline packages/gatsby/src/commands/serve.ts:115]: are we dropping the flag now?

Comment (KyleAMathews) [reply packages/gatsby/src/commands/serve.ts:115]: oh hmm no y et — must have accidentally deleted that while refactoring. Restoring.

**Review by dmccraw — DISMISSED** [review]
:ship: it! It looks good to me.

**Review by dmccraw — DISMISSED** [review]
:ship: it!

File: packages/gatsby/src/commands/serve.ts (hunk 1 above)
Comment (KyleAMathews) [reply packages/gatsby/src/commands/serve.ts:115]: ugh, I did mean to do this — `gatsby serve` doesn't have access to flags in gatsby-config.js so just looking for the functions manifest is the best way to see if it's enabled or not

**Review by dmccraw — APPROVED** [review]
:ship: it!

**Pull request merged**